src/GraphQL/GraphQLExtensions.cs

```diff
@@ -483,5 +485,91 @@ public static IValue AstFromValue(this object value, ISchema schema, IGraphType
         }
+
+        /// <summary>
+        /// Add <paramref name="directive"/> to object metadata and apply it.
+        /// </summary>
+        /// <param name="metadataProvider"> Metadata. </param>
+        /// <param name="name"> Directive name. </param>
+        /// <param name="directive"> Schema directive. </param>
+        public static void AddDirective(this IProvideMetadata metadataProvider, string name, SchemaDirectiveVisitor directive)
+        {
+            // save directive to metadata
+            metadataProvider.SetDirective(name, directive);
+
+            // apply directive to target
+            switch (metadataProvider)
+            {
+                case Schema schema:
+                    directive.VisitSchema(schema);
+                    break;
+                case ObjectGraphType objectGraphType:
+                    directive.VisitObjectGraphType(objectGraphType);
+                    break;
+                case IObjectGraphType objectGraphType:
+                    directive.VisitObjectGraphType(objectGraphType);
+                    break;
+                case EnumerationGraphType enumeration:
+                    directive.VisitEnum(enumeration);
+                    break;
+                case EnumValueDefinition value:
+                    directive.VisitEnumValue(value);
+                    break;
+                case ScalarGraphType scalar:
+                    directive.VisitScalar(scalar);
+                    break;
+                case FieldType field:
+                    directive.VisitField(field);
+                    break;
+                case QueryArgument argument:
+                    directive.VisitArgumentDefinition(argument);
+                    break;
+                case InterfaceGraphType interfaceGraphType:
+                    directive.VisitInterface(interfaceGraphType);
+                    break;
+                case UnionGraphType union:
+                    directive.VisitUnion(union);
+                    break;
+                case InputObjectGraphType type:
+                    directive.VisitInputObject(type);
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Add <paramref name="directive"/> to object metadata.
+        /// </summary>
+        /// <param name="metadataProvider"> Metadata. </param>
+        /// <param name="name"> Directive name. </param>
+        /// <param name="directive"> Schema directive. </param>
+        public static void SetDirective(this IProvideMetadata metadataProvider, string name, SchemaDirectiveVisitor directive)
+        {
+            Dictionary<string, SchemaDirectiveVisitor> directives;
+            if (metadataProvider.HasMetadata(SchemaDirectivesMetadataKey))
+            {
+                directives = metadataProvider.GetMetadata(SchemaDirectivesMetadataKey, new Dictionary<string, SchemaDirectiveVisitor>());
```

Comment:
So if we are here than there is no need to allocate default value `new Dictionary<string, SchemaDirectiveVisitor>()` since metadata is not empty.

Reply:
I looked more closely and realized that the second parameter is convenient for inferring the type of result. But this requires an instance to be created, which creates unnecessary memory allocation if the necessary metadata is available. So I suggest to add `TType IProvideMetadata.GetMetadata<TType>(string key, Func<TType> defaultValueFactory);` with that implementation:
```c#
public TType GetMetadata<TType>(string key, Func<TType> defaultValueFactory)
        {
            var local = Metadata;
            return local != null && local.TryGetValue(key, out var item) ? (TType)item : defaultValueFactory();
        }
```
and the caller will look like this
```c#
directives = metadataProvider.GetMetadata(SchemaDirectivesMetadataKey, () => new Dictionary<string, SchemaDirectiveVisitor>());
```

Just add `()=>` and no more unnecessary allocations :)

Reply:
Replace with `metadataProvider.GetMetadata<Dictionary<string, SchemaDirectiveVisitor>>(SchemaDirectivesMetadataKey)`

Reply:
Extended `IProvideMetadata` and inherited all possible classes from default implementation `MetadataProvider` to remove identical implementation

Reply:
I think such a change with new base class should go as different PR. This will allow you to break down the issue into small separate problems and quickly merge it into master

Reply:
Ok, create separate PR - https://github.com/graphql-dotnet/graphql-dotnet/pull/1349